src/v/v8_engine/executor.h

```diff
@@ -0,0 +1,210 @@
+/*
+ * Copyright 2021 Vectorized, Inc.
+ *
+ * Licensed as a Redpanda Enterprise file under the Redpanda Community
+ * License (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * https://github.com/vectorizedio/redpanda/blob/master/licenses/rcl.md
+ */
+
+#pragma once
+
+#include "seastarx.h"
+
+#include <seastar/core/abort_source.hh>
+#include <seastar/core/future.hh>
+#include <seastar/core/gate.hh>
+#include <seastar/core/lowres_clock.hh>
+#include <seastar/core/reactor.hh>
+#include <seastar/core/sharded.hh>
+
+#include <boost/lockfree/queue.hpp>
+
+#include <atomic>
+#include <condition_variable>
+#include <memory>
+#include <semaphore.h>
+
+namespace v8_engine {
+
+namespace internal {
+
+// This class is used for notify seastar thread when executor thread will
+// complete task
+class condition {
+public:
+    condition();
+
+    // Run this function in seastar thread.
+    ss::future<> wait();
+
+    // Run this function in executor thread
+    void notify();
+
+private:
+    ss::file_desc _file_desc;
+    int _fd;
+    ss::pollable_fd _pollable_fd;
+    eventfd_t _event{0};
+};
+
+// Abstract class for executor task
+struct work_item {
+    virtual ~work_item() {}
+    virtual void process() = 0;
+};
+
+// This class implement task for executor. Contains promise logic
+template<typename Func>
+struct task final : work_item {
+    explicit task(Func&& f)
+      : _func(std::move(f)) {}
+
+    // Process task from executor in executor thread and set state in ss::future
+    void process() override {
+        try {
+            _func();
+            _state.set(true);
+        } catch (...) {
+            _state.set_exception(std::current_exception());
+        }
+        _on_done.notify();
+    }
+
+    // Get future for seastar thread for wating when executor thread will
+    // complete task
+    ss::future<> get_future() {
+        return _on_done.wait().then([this] {
+            if (_state.failed()) {
+                return ss::make_exception_future<>(
+                  std::move(_state).get_exception());
+            } else {
+                return ss::make_ready_future<>();
+            }
+        });
+    }
+
+    Func _func;
+    ss::future_state<bool> _state;
+    condition _on_done;
+};
+
+// This class implement wrapper for ss::gate for using in ss::sharded
+class local_gate {
```

Comment:
same deal here, you won't need this class if you just want a sharded<ss::gate> you can do that

Reply:
done